components/locid/src/subtags/region.rs

```diff
@@ -41,22 +41,25 @@ impl Region {
     ///
     /// assert_eq!(region, "FR");
     /// ```
-    pub fn from_bytes(v: &[u8]) -> Result<Self, ParserError> {
-        match v.len() {
-            REGION_ALPHA_LENGTH => {
-                let s = TinyAsciiStr::from_bytes(v).map_err(|_| ParserError::InvalidSubtag)?;
-                if !s.is_ascii_alphabetic() {
-                    return Err(ParserError::InvalidSubtag);
-                }
+    pub const fn from_bytes(v: &[u8]) -> Result<Self, ParserError> {
+        Self::from_bytes_manual_slice(v, 0, v.len())
+    }
+
+    /// Equivalent to [`from_bytes(bytes[start..end])`](Self::from_bytes),
+    /// but callable in a `const` context (which range indexing is not).
+    pub const fn from_bytes_manual_slice(
+        v: &[u8],
+        start: usize,
+        end: usize,
+    ) -> Result<Self, ParserError> {
+        match (
+            end - start,
+            TinyAsciiStr::from_bytes_manual_slice(v, start, end),
+        ) {
```

Comment:
`question`: this should cost perf. You are forcing tinystr construction for each attempt to test if the tag is a region. Before that, the code is heavily optimized to bail with as few operations as possible ordered from the most likely reasons to bail to the least likely.
This allows consumers to write all their code with `if let Ok(region) = Region::try_from_x()` quite abundantly.

I think that since multiple subtags can match the same string, it is a good practice to keep it, over separating `does_slice_match` vs `from_slice` for testing/constructing.

Can you look back at the from slice constructors that you rewrote and shift from optimizing for code conscience to optimize for cheap bailouts and nest more to only pay the cost when needed? (e.g. testing length is cheap, so can be done early, but construction of tinystr is not, so do it only if length matches etc.).

Reply:
(I am making an assumption above that rust match for tuples is not smart enough to delay second item in the tuple until first one matches, but I didn't check the assembly. Feel free to evaluate)